### ISSUE TITLE
Bug in loading structured array

### ISSUE THREAD

**Issue by dmbelov** [Blosc/bloscpack#16]
Bloscpack  version 0.7.1 incorrectly stores or loads metadata for a structured Numpy array.  Could you please correct it?

Here is an example:
```python
import bloscpack
arr = numpy.array([('a', 1), ('b', 2)], dtype=[('a', 'S1'), ('b', 'f8')])

bloscpack.pack_ndarray_file(arr, 'arr.blp')
bloscpack.unpack_ndarray_file('arr.blp')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[I removed most of the traceback except the last call]

/usr/bin/anaconda/lib/python2.7/site-packages/bloscpack/numpy_io.py in __init__(self, metadata)
     82             raise NotANumpyArray
     83 	    self.ndarray = numpy.empty(metadata['shape'],
---> 84                                    dtype=numpy.dtype(metadata['dtype']),
     85                                    order=metadata['order'])
     86         self.ptr = self.ndarray.__array_interface__['data'][0]

TypeError: data type not understood
```

In the debugger we clearly see the problem --- numpy does NOT understand the format of `metadata['dtype']`:
```python
ipdb> metadata['dtype']
[[u'a', u'|S1'], [u'b', u'<f8']]
```
Note that:
 1. the inner list must be a tuple;
 2. string must be usual strings rather than unicode ones.

To make it working, I corrected the code on lines 83-86 in `numpy_io.py` as follows, but, probably, there is a better solution:
```python
/usr/bin/anaconda/lib/python2.7/site-packages/bloscpack/numpy_io.py in __init__(self, metadata)

        def _conv(descr):
            """
           Converts nested list of lists into list of tuples. Examples::
             [[u'a', u'f8']] -> [('a', 'f8')]
             [[u'a', u'f8', 2]] -> [('a', 'f8', 2)]
             [[u'a', [[u'b', 'f8']]]] -> [('a', [('b', 'f8')])]
            """
            if isinstance(descr, list):
                if isinstance(descr[0], list):
                    descr = map(_conv, descr)
                else:
                    descr = tuple(map(_conv, descr))
            elif isinstance(descr, unicode):
                descr = str(descr)
            else:
                # keep descr as is
                pass
            return descr

        self.ndarray = numpy.empty(metadata['shape'],
                                   dtype=numpy.dtype(_conv(metadata['dtype'])),
                                   order=metadata['order'])
```

Thanks,
Dmitry.

**Comment by esc** [comment]
I am hunching this may be resolved soon.

**Comment by esc** [comment]
Solution in #24

**Comment by esc** [comment]
I have fixed this, but I fear that it will cause backwards incompatability if people already have existing code.

**Comment by esc** [comment]
Backwards compat enabled with #33

**Comment by esc** [comment]
@dmbelov this was included in the new release, did you ever get a chance to see if it works for you?

**Comment by dmbelov** [comment]
I updated to new version.  However it is NOT backward compatible --- I cannot read my old blp files.   The problem is in this code in numpy_io.py:

        try:
            dtype_ = numpy.safe_eval(metadata['dtype'])
        except SyntaxError:
            dtype_ = metadata['dtype']

numpy.safe_eval raises TypeError exception and the value of metadata['dtype'] in my case is
[[u'current_ric', u'|S8'], [u'price', u'<f8'], [u'volume', u'<f8'], [u'datetime', u'<M8[ms]']]

What do you suggest to do?

**Comment by dmbelov** [comment]
To overcome incompatibility issue I added the following logic in my local checkout:

```
def _conv(descr):
    """
    Converts nested list of lists into list of tuples. Examples::
     [[u'a', u'f8']] -> [('a', 'f8')]
     [[u'a', u'f8', 2]] -> [('a', 'f8', 2)]
     [[u'a', [[u'b', 'f8']]]] -> [('a', [('b', 'f8')])]
    """
    if isinstance(descr, list):
        if isinstance(descr[0], list):
            descr = map(_conv, descr)
        else:
            descr = tuple(map(_conv, descr))
    elif isinstance(descr, unicode):
        descr = str(descr)
    else:
        # keep descr as is
        pass
    return descr

class PlainNumpySink(PlainSink):
       [...]

        try:
            dtype_ = numpy.safe_eval(metadata['dtype'])
        except SyntaxError:
            dtype_ = metadata['dtype']
        except TypeError, exc:
            dtype_ = _conv(metadata['dtype'])
```

**Comment by esc** [comment]
First of all, thanks for checking and getting back to me.

The backwards compatibility only works for non-nested dtypes, i.e. for everything that could have been saved AND loaded. The dtype you describe was saved in a format that was never intended be successfully re-loaded, that was the bug. The fix consisted of saving such dtypes in a better format, namely one that can be roundtripped through JSON (and probably also other serializers).

I am not sure what to do. Can you use your fix to re-pack all the existing data-files you have? Or do you have too many? If you have too many I'll consider adding the hack for data that had been saved but couldn't be loaded, like yours.

**Comment by dmbelov** [comment]
I have few 1000 files like that...  It would be great if you add a code for backward compatibility.

**Comment by esc** [comment]
Oh, ok, didn't know you were such a heavy user. I'll add backwards compat ASAP. Will it be enough for the time being to just add the fix you made above? Do you think there are poetntially other numpy dtypes that might not be loadable with the fix?

**Comment by dmbelov** [comment]
Thanks for this.
There is no rash: I've already inserted that hack into my local copy of bloscpack.  So, I do not have any problems with loading files.

I think that the function _conv covers all cases that are produced by the previous "buggy" version of numpy_io.  The only question is whether you need to call "_conv" in "except TypeError" or there is a better way for checking that metadata['dtype'] is in the old format.

**Comment by esc** [comment]
A format version would have been useful...

**Comment by FrancescAlted** [comment]
Better introducing it now than never:)

Enviat des del meu iPhone

El 17/04/2015, a les 19:06, Valentin Haenel <notifications@github.com> va escriure:

**Comment by esc** [comment]
Can you implement a test case that exposes the issue against ``0.7.x-maint``?

**Comment by esc** [comment]
@dmbelov I just released v0.7.3 , now available on PyPi which should fix exactly that issue.
